Channel mentioning rewrite + refactor + add support for suggesting emotes and emojis

saltrafael opened the pull request:
🟢    https://salt.odysee.com

## Fixes

Issue Number: Closes #68 Closes #135 Closes #256

- Moved channel mentioning away from commentCreate
  - now works in any textarea field (can be disabled where needed)
- Replaced reach/ui combobox for mui/Autocomplete
 - Added support for suggesting emojis when typing :

## PR Checklist

<!-- For the checkbox formatting to work properly, make sure there are no spaces on either side of the "x" -->

<details><summary>Toggle...</summary>

What kind of change does this PR introduce?

- [X] Bugfix
- [X] Feature
- [X] Code style update (formatting)
- [X] Refactoring (no functional changes)
- [ ] Documentation changes
- [ ] Other - Please describe:
</details>

dan-peterson commented on ui/component/collectionEdit/view.jsx:
```diff
@@ -1,112 +1,113 @@
   const [params, setParams]: [any, (any) => void] = React.useState({});
+
+  const activeChannelName = activeChannelClaim && activeChannelClaim.name;
+  let prefix = IS_WEB ? `${DOMAIN}/` : 'lbry://';
```
Do we still need to check IS_WEB?

infinite-persistence commented on ui/component/commentCreate/view.jsx:
```diff
@@ -11,11 +11,9 @@ import * as ICONS from 'constants/icons';
 import * as KEYCODES from 'constants/keycodes';
```
After typing some comments in livestream, the HOME and END key does not bring the cursor to the start and end of the line.  Reverted the PR, then it'll work.

infinite-persistence reviewed (changes requested):
## Review Part 1 (~40% done)
Still going through the changes, but thought of passing along the findings first.

<details>
  <summary>Also, some notes for future...</summary>
  
1. Try to keep 1 commit per change, e.g. "moved xxx to yyy", "re-ordered parameters", "consolidate/DRY parseName", "remove desktop code". This makes the delta consumable for review and for future reference when we need to do partial revert.
2. Re-ordered code is the hardest to review. Restrict those into single-commits with "no functional change" in the commit message, so that it'll be easier to review (we can then skim/skip these commits by assuming no new functionality was added). Make functional changes in the next commit instead.
3. I think we should stop re-ordering import lines until we come to a consensus on how it should be ordered (we can start a new thread in Slack).  I believe Sean's ordering was:  constants -> external (React) -> internal; while the one in this PR is a direct line sort.  I think line sorts don't work well because of `{ }`, and we'll end up with unnecessary deltas in the future.
    - e.g.  If we change from `import { blah }` to `import React, { blah }`, it will involve a re-ordering and a less-readable delta.
    - There might be plugins to auto-sort at commit time based on agreed rules (like our linter), so that we aren't fighting each other's style + no need to manually sort.

</details>

infinite-persistence commented on ui/component/textareaWithSuggestions/view.jsx:
```diff
@@ -0,0 +1,330 @@
+  /** Functions **/
+  /** --------- **/
+
+  function handleChange(e: SyntheticInputEvent<*>) {
```
I typed "hello world :angry @" 
On the "@", I was expecting channel suggestions to appear, but it stayed at emote suggestion.

saltrafael commented:
`REBASE`

moved out the regex changes which made it too big and out of scope

`DRAFT`

have more changes to do, fix the requested changes

infinite-persistence commented on ui/component/common/form-components/form-field.jsx:
```diff
@@ -4,52 +4,59 @@ import { FF_MAX_CHARS_DEFAULT } from 'constants/form-field';
 import React from 'react';
 import ReactDOMServer from 'react-dom/server';
 import SimpleMDE from 'react-simplemde-editor';
+import TextareaWithSuggestions from 'component/textareaWithSuggestions';
```
Underscore, for example ":angry_" , dismisses the suggestion popup.  It should still be there since `:angry_1` is a selection.

infinite-persistence commented on ui/component/textareaWithSuggestions/index.js:
```diff
@@ -1,27 +1,32 @@
+import TextareaWithSuggestions from './view';
 
 const select = (state, props) => {
+  const claim = makeSelectClaimForUri(props.uri)(state);
```
Please help to start switching to `selectClaimForUri(state, props.uri)` (note the change in function signature as well).

infinite-persistence commented:
> _moved out the regex changes which made it too big and out of scope_
It's ok if you want to leave it there.  I've reviewed the first 3 commits ... only last 2 commits left to review.
The separation note is more for the future :)

saltrafael commented:
- Rebased master
- Improved group label to display the matching term (similar to discord)
- Added back support for searching mention
- Added support for typing emojis (it works very similar to slack for comparison)

infinite-persistence commented on ui/component/textareaWithSuggestions/view.jsx:
```diff
@@ -0,0 +1,387 @@
+  };
+
+  return (
+    <Autocomplete
```
After typing `@` in a comment text field.  Just need to put a key for each item in the list.

![image](https://user-images.githubusercontent.com/64950861/144954845-c33d24fc-c383-4d71-b11b-1f523e09dcfe.png)

infinite-persistence commented on ui/component/textareaWithSuggestions/view.jsx:
```diff
@@ -0,0 +1,387 @@
+
+  return (
+    <Autocomplete
+      autoHighlight
```
Also from `@` in the text field:

![image](https://user-images.githubusercontent.com/64950861/144955127-f873e90f-13b0-46c2-9297-131302be80a1.png)

infinite-persistence commented on ui/component/textareaWithSuggestions/view.jsx:
```diff
@@ -0,0 +1,387 @@
+      id={id}
+      inputValue={messageValue}
+      loading={!allMatches || allMatches.length === 0 || showPlaceholder}
+      loadingText={results || showPlaceholder ? __('Searching...') : __('Nothing found')}
```
"Searching..." in invisible in dark mode.

![image](https://user-images.githubusercontent.com/64950861/144955244-72e601b5-9393-4b4d-8a3e-7a4c6b709682.png)

infinite-persistence commented on ui/component/textareaWithSuggestions/view.jsx:
```diff
@@ -0,0 +1,387 @@
+      inputValue={messageValue}
+      loading={!allMatches || allMatches.length === 0 || showPlaceholder}
+      loadingText={results || showPlaceholder ? __('Searching...') : __('Nothing found')}
+      onBlur={() => onBlur && onBlur()}
```
In this video, the search results came back, but the gui is stuck at "Searching..."

https://www.notion.so/Temp-Links-31f1387c619d4f338b877c41b99fd59b#b1662ab22da947acbdd79f7350f8a08a

infinite-persistence reviewed (changes requested):
Initial glance.
Looks smooth.

saltrafael commented on ui/component/textareaWithSuggestions/view.jsx:
```diff
@@ -0,0 +1,387 @@
+  const renderGroup = (group: string, children: any) => (
+    <div className="textareaSuggestions__group">
+      <label className="textareaSuggestions__label">
+        {group}
+        {suggestionTerm && suggestionTerm.length > 1
+          ? ' ' + __('matching %matching_term%', { matching_term: suggestionTerm })
+          : ''}
```
good find!

saltrafael commented on package.json:
```diff
@@ -48,6 +48,9 @@
   "dependencies": {
+    "@emotion/react": "^11.6.0",
+    "@emotion/styled": "^11.6.0",
+    "@mui/material": "^5.2.1",
```
I had the mui switch explanation saved as draft on slack and forgot to send 😆 ! it is there now for discussion

infinite-persistence commented:
Glanced through latest changes ... LGTM.
Didn't run and try it though.

tzarebczan commented:
Yep, seems to be working well with initial testing outside of the livestream comments. Once that's fixed, will ship.
